libraries/AP_FETtecOneWire/AP_FETtecOneWire.h

```diff
@@ -0,0 +1,253 @@
+        @param length length of the bytes array
+    */
+    void transmit(const uint8_t esc_id, const uint8_t *bytes, uint8_t length);
+
+    /**
+        reads the FETtec OneWire answer frame of an ESC
+        @param bytes 8 bit byte array, where the received answer gets stored in
+        @param length the expected answer length
+        @param return_full_frame can be return_type::RESPONSE or return_type::FULL_FRAME
+        @return 2 on CRC error, 1 if the expected answer frame was there, 0 if dont
+    */
+    receive_response receive(uint8_t *bytes, uint8_t length, return_type return_full_frame);
+
+    /**
+        Resets a pending pull request
+    */
+    void pull_reset();
+
+    /**
+        Pulls a complete request between flight controller and ESC
+        @param esc_id id of the ESC
+        @param command 8bit array containing the command that should be send including the possible payload
+        @param response 8bit array where the response will be stored in
+        @param return_full_frame can be return_type::RESPONSE or return_type::FULL_FRAME
+        @param req_len transmit request length
+        @return true if the request is completed, false if dont
+    */
+    bool pull_command(const uint8_t esc_id, const uint8_t *command, uint8_t *response, return_type return_full_frame, const uint8_t req_len);
+
+    /**
+        Scans for all ESCs in bus. Configures fast-throttle and telemetry for the ones found.
+        Should be periodically called until _scan.state == scan_state_t::DONE
+    */
+    void scan_escs();
+
+    /**
+        configure the fast-throttle command.
+        Should be called once after scan_escs() is completted and before config_escs()
+    */
+    void config_fast_throttle();
+
+#if HAL_WITH_ESC_TELEM
+    /**
+        increment message packet count for every ESC
+    */
+    void inc_send_msg_count();
+
+    /**
+        calculates tx (outgoing packets) error-rate by converting the CRC error counts reported by the ESCs into percentage
+        @param esc_id id of ESC, that the error is calculated for
+        @param esc_error_count the error count given by the esc
+        @return the error in percent
+    */
+    float calc_tx_crc_error_perc(const uint8_t esc_id, uint16_t esc_error_count);
+
+    /**
+        if init is complete checks if the requested telemetry is available.
+        @param t telemetry datastructure where the read telemetry will be stored in.
+        @param centi_erpm 16bit centi-eRPM value returned from the ESC
+        @param tx_err_count Ardupilot->ESC communication CRC error counter
+        @param tlm_from_id receives the ID from the ESC that has respond with its telemetry
+        @return 1 if CRC is correct, 2 on CRC mismatch, 0 on waiting for answer
+    */
+    receive_response decode_single_esc_telemetry(TelemetryData& t, int16_t& centi_erpm, uint16_t& tx_err_count, uint8_t &tlm_from_id);

```

Comment:
is centi_erpm ever a negative number? I couldn't see. if not, it should be unsigned, ie uint16_t  ... in a number of places.

Reply:
These ESCs can reverse rotation, so I guess it can be

Reply:
We now support reversing direction, so this can be tested now.